### ISSUE TITLE
🤗 [Question]: How to stream large file for download ?

### ISSUE THREAD

**Issue by sgaunet** [gofiber/fiber#2429]
### Question Description

 I would like to stream the content of a file from s3 without downloading it on the local storage. I would like to do all in memory with buffering. I'm able to do it for the upload but I'm stuck for the download.

So I can't use Sendfile or Download functions.  I've tried to use Sendstream without any success.

What solutions I have to achieve that ?

### Checklist:

- [X] I agree to follow Fiber's [Code of Conduct](https://github.com/gofiber/fiber/blob/master/.github/CODE_OF_CONDUCT.md).
- [X] I have checked for existing issues that describe my questions prior to opening this one.
- [X] I understand that improperly formatted questions may be closed without explanation.

**Comment by gaby** [comment]
Adding a comment to keep track of this. I have to check if this is supported by Fashttp using Streaming.

**Comment by gaby** [comment]
I will take a look at this during the weekend

**Comment by gaby** [comment]
@tomaswarynyca I haven't tested if its possible to access this from Fiber. It might need code changes.

There's a new boolean that would need to be added `StreamResponseBody` and the response stream becomes available via the context

**Comment by codemicro** [comment]
If I am understanding your issue correctly @tomaswarynyca, this might be helpful. It's a modified version of [this example written for FastHTTP](https://github.com/valyala/fasthttp/issues/422#issuecomment-422387236) to stream a response from the server.

```go
package main

import (
	"bufio"
	"fmt"
	"github.com/gofiber/fiber/v2"
	"time"
)

func main() {
	app := fiber.New()

	app.Get("/source", func(ctx *fiber.Ctx) error {
		ctx.Set("Content-Type", "text/event-stream")
		fr := ctx.Response()
		fr.SetBodyStreamWriter(func(w *bufio.Writer) {
			for {
				w.Write([]byte(fmt.Sprintf("data: %s\n\n", time.Now())))

				if err := w.Flush(); err != nil {
					println("client disconnected")
					return
				}

				time.Sleep(time.Second)
			}
		})
		return nil
	})

	app.Get("/", func(ctx *fiber.Ctx) error {
		ctx.Type("html")
		return ctx.SendString(page)
	})

	app.Listen("127.0.0.1:9090")
}

const page = `<!doctype html>
<body>
<div id=d></div>
<script>
var source = new EventSource('http://127.0.0.1:9090/source');

source.onmessage = function(e) {
  document.getElementById('d').innerHTML = e.data;
};
</script>`
```

**Comment by gaby** [comment]
@tomaswarynyca Did the example work for you?

**Comment by vorticalbox** [comment]
I have only been playing with go a few days so not well versed in the state of aws sdk (i am mainly a nodejs dev) but a better way would be to get the client to down the file directly from s3. 

the flow goes like this.

* Client sends request to api for download a file
* your api generates a get object signed url for the file from the aws sdk
* your api then sends a http redirect to the signed url

This will also save you server load as the file does not stream through your app.

**Comment by gaby** [comment]
> I have only been playing with go a few days so not well versed in the state of aws sdk (i am mainly a nodejs dev) but a better way would be to get the client to down the file directly from s3.
> 
> the flow goes like this.
> 
> * Client sends request to api for download a file
> * your api generates a get object signed url for the file from the aws sdk
> * your api then sends a http redirect to the signed url
> 
> This will also save you server load as the file does not stream through your app.

This is not related to Fiber, you can ask in the "Go Support" chat of our Discord.

**Comment by vorticalbox** [comment]
> > I have only been playing with go a few days so not well versed in the state of aws sdk (i am mainly a nodejs dev) but a better way would be to get the client to down the file directly from s3.
> > the flow goes like this.
> > 
> > * Client sends request to api for download a file
> > * your api generates a get object signed url for the file from the aws sdk
> > * your api then sends a http redirect to the signed url
> > 
> > This will also save you server load as the file does not stream through your app.
> 
> This is not related to Fiber, you can ask in the "Go Support" chat of our Discord.

I know I was just trying to suggest a better way than stream large file through an api when you didn't have to.

**Comment by peczenyj** [comment]
Some tips about handle large files over http

1. If you are using an nginx as proxy to your app, you should consider use it to serve the (big) static files IF APPLICABLE 
2. If you can use a CDN, use it
3. responding to several big file requests can be some kind of DDoS, think about the risks and use some limiter (by ip? ) if possible 
4. Consider support range
5. Check if you can take some advantage of DSR (direct server response) in your network

**Comment by mlycore** [comment]
I ran into a similar problem while using Fiber middleware proxy to download large files, and this will cause a sudden memory increase. After profiling, I found it Fiber had read all backend data into memory and then send it back to the client. I tried set StreamBody but didn't work.

**Comment by gaby** [comment]
@mlycore Correct, for streaming data you can't use the middleware. I'm not sure if it may be possible to add a method for doing that.

We would have to create an HTTP client that takes an `io.Reader`

**Comment by mlycore** [comment]
> @mlycore Correct, for streaming data you can't use the proxy middleware. I'm not sure if it may be possible to add a method for doing that.
> 
> We would have to create an HTTP client that takes an `io.Reader`

For pprof, it's fasthttp's method `appendBodyFixedSize()` which consumed that much memory. 

![image](https://github.com/gofiber/fiber/assets/22319610/41164f46-84e8-4727-89ef-4512d61fe838)

**Comment by codemicro** [comment]
> For pprof, it's fasthttp's method appendBodyFixedSize() which consumed that much memory.

@mlycore if you want to stream data using Fiber, you can modify [this example further up in the thread](https://github.com/gofiber/fiber/issues/2429#issuecomment-1571210179) to suit your needs - Fiber has no API for *actually* streaming data (the `(*Ctx).SendStream` method actually just reads everything out of your `io.Reader` and sends that as one big blob), you have to delve into the FastHTTP interface to get it to work.

**Comment by gaby** [comment]
> > For pprof, it's fasthttp's method appendBodyFixedSize() which consumed that much memory.
> 
> @mlycore if you want to stream data using Fiber, you can modify [this example further up in the thread](https://github.com/gofiber/fiber/issues/2429#issuecomment-1571210179) to suit your needs - Fiber has no API for _actually_ streaming data, you have to delve into the FastHTTP interface to get it to work. The `(*Ctx).SendStream` method actually just reads everything out of your `io.Reader` and sends that as one big blob, which explains the memory usage you're seeing.

That's incorrect. You can Stream a Request Body, you can't Stream a Response (yet even though fasthttp supports it). 

Most use cases are about Stream Upload which Fiber already supports.

**Comment by codemicro** [comment]
> You can Stream a Request Body, you can't Stream a Response (yet even though fasthttp supports it).

Admittedly, my answer was based on the assumption that they're streaming a response *from* their application based on the context of the rest of the thread.

**Comment by kawache** [comment]
if you are behind an nginx, you can you use Send file feature, basically you can have your business logic like checking for user's permission to access a large file. and then return a http response including the url on of the file ( like is protected ) in [x-accel.redirect](https://www.nginx.com/resources/wiki/start/topics/examples/x-accel/#x-accel-redirect) header to be intercepted by Nginx, who will take care of reading and serving the file.

**Comment by gaby** [comment]
@mlycore @codemicro This is related to https://github.com/valyala/fasthttp/issues/2037

**Comment by ReneWerner87** [comment]
v3 covers this. `c.SendStream(r io.Reader, size ...int)` streams a reader straight to the response without buffering it, and `c.SendStreamWriter(func(*bufio.Writer))` hands you the writer for chunked output.

```go
app.Get("/download", func(c fiber.Ctx) error {
    obj, err := s3Client.GetObject(c.Context(), input)
    if err != nil {
        return err
    }
    defer obj.Body.Close()

    c.Set(fiber.HeaderContentDisposition, `attachment; filename="file.bin"`)
    return c.SendStream(obj.Body) // pass the size as the second argument when known
})
```

Omit the size and fasthttp streams the body chunked; pass it and you get a `Content-Length`.

Closing.
